kotlinx-coroutines-core/common/src/flow/operators/Share.kt

```diff
@@ -0,0 +1,400 @@
+ *     .shareIn(scope, 0)
+ * ```
+ *
+ * Any exception in the upstream flow terminates the sharing coroutine without affecting any of the subscribers,
+ * and will be handled by the [scope] in which the sharing coroutine is launched. Custom exception handling
+ * can be configured by using the [catch] or [retry] operators before the `shareIn` operator.
+ * For example, to retry connection on any `IOException` with 1 second delay between attempts, use:
+ *
+ * ```
+ * val messages = backendMessages
+ *     .retry { e ->
+ *         val shallRetry = e is IOException // other exception are bugs - handle them
+ *         if (shallRetry) delay(1000)
+ *         shallRetry
+ *     }
+ *     .shareIn(scope, 0)
+ * ```
+ *
+ * ### Initial value
+ *
+ * When a special initial value is needed to signal to subscribers that the upstream is still loading the data,
+ * use the [onStart] operator on the upstream flow. For example:
+ *
+ * ```
+ * backendMessages
+ *     .onStart { emit(UpstreamIsStartingMessage) }
+ *     .shareIn(scope, 1) // replay one most recent message
+ * ```
+ *
+ * ### Buffering and conflation
+ *
+ * The `shareIn` operator runs the upstream flow in a separate coroutine, and buffers emissions from upstream as explained
+ * in the [buffer] operator's description, using a buffer of [replay] size or the default (whichever is larger).
+ * This default buffering can be overridden with an explicit buffer configuration by preceding the `shareIn` call
+ * with [buffer] or [conflate], for example:
+ *
+ * * `buffer(0).shareIn(scope, 0)` &mdash; overrides the default buffer size and creates a [SharedFlow] without a buffer.
+ *   Effectively, it configures sequential processing between the upstream emitter and subscribers,
+ *   as the emitter is suspended until all subscribers process the value. Note, that the value is still immediately
+ *   discarded when there are no subscribers.
+ * * `buffer(b).shareIn(scope, r)` &mdash; creates a [SharedFlow] with `replay = r` and `extraBufferCapacity = b`.
+ * * `conflate().shareIn(scope, r)` &mdash; creates a [SharedFlow] with `replay = r`, `onBufferOverflow = DROP_OLDEST`,
+ *   and `extraBufferCapacity = 1` when `replay == 0` to support this strategy.
+ *
+ * ### Operator fusion
+ *
+ * Application of [flowOn][Flow.flowOn], [buffer] with [RENDEZVOUS][Channel.RENDEZVOUS] capacity,
+ * or [cancellable] operators to the resulting shared flow has no effect.
+ *
+ * ### Exceptions
+ *
+ * This function throws [IllegalArgumentException] on unsupported values of parameters or combinations thereof.
+ *
+ * @param scope the coroutine scope in which sharing is started.
+ * @param replay the number of values replayed to new subscribers (cannot be negative).
+ * @param started the strategy that controls when sharing is started and stopped.
+ */
+@ExperimentalCoroutinesApi
+public fun <T> Flow<T>.shareIn(
+    scope: CoroutineScope,
+    replay: Int,
```

Comment:
Why doesn't `replay` default to `0`? I find it quite confusing to see zero everywhere and could be further confusing with flows of numbers.
In the following snippet (taken from replaceWith clauses in this PR), do you find it obvious what `0` is when you don't know by heart what it is?
```
shareIn(scope, 0, SharingStarted.Lazily)
```
I believe replay should be disabled by default (so only the latest value is collected by collectors while the flow is active), which would translate to a default value of zero, and snippets should encourage to specify `replay = ` or `replayCount = ` for better explicitedness.

Reply:
We might introduce defaults later on in future updates. For now, we leave it fully explicit, so that it is easier for us to analyze how people actually use it.

Reply:
It seems I'm not the only one to be a little puzzled when I see the examples of the `0` argument without any name in the snippets from this PR: https://github.com/Kotlin/kotlinx.coroutines/issues/2047#issuecomment-707370301

Reply:
Thanks a lot for noticing it. Now I actually see a problem that I cannot unsee &mdash; the parameters are in incosonsitent order between `stateIn` and `shareIn` operator. Will fix and add default, too.

Reply:
I've seen the changes, great!
I like the new order better, I find it somehow more familiar to have `replay` in the last position after the scope, and after the strategy.